### PR TITLE
Faceting

### PR DESCRIPTION
**Pull request by Hodgy**
**Shouldn't be merged until #7 has been merged as this is based off that branch**

Adds search facets.
We just provide the facet data to the theme to present to the user.
Does feature some hardcoded defaults for an internal project. The will be removed for a changeable option via the settings page

### CODE REVIEW

File: includes/class-wp-lucene-search-abstract-searcher.php [hunk 1]
```diff
@@ -22,9 +24,10 @@ public function search($search_term)
 
     /**
      * @param $search
+     * @param $args
      * @return mixed
```
Comment (cooperaj) [inline includes/class-wp-lucene-search-abstract-searcher.php]: fix yo docblocks

File: includes/class-wp-lucene-search-helper.php [hunk 2]
```diff
@@ -83,4 +83,59 @@ protected static function valid_search_service($search_service)
...
+
+        if (empty( $args )) {
+            return '';
+        }
+
+        $string = '';
+        $facets = [];
+        foreach ($args as $field => $logic) {
+            foreach ( $logic as $key => $value ) {
+                $string .= 'wls['.$field .  '][' . $key . ']='. $value . '&';
```
Comment (cooperaj) [inline includes/class-wp-lucene-search-helper.php]: Can we split into other methods?

File: includes/searcher/class-wp-lucene-search-searcher-azure.php [hunk 3]
```diff
@@ -13,13 +13,18 @@ protected function build_url()
         return $url;
     }
 
-    public function build_body($search)
+    public function build_body($search, $args)
     {
+        $facets =  array_key_exists( 'facets', $args) ? $this->build_facets($args['facets']) : [];
```
Comment (cooperaj) [inline includes/searcher/class-wp-lucene-search-searcher-azure.php]: Brackets

File: includes/searcher/class-wp-lucene-search-searcher-azure.php [hunk 4]
```diff
@@ -28,15 +33,76 @@ protected function get_request_object( $body )
         return new WP_Lucene_Search_Request_Azure( $this->build_url(), json_encode( $body ) );
     }
 
-    protected function parse_result($result)
+    protected function parse_result( $result )
```
Comment (cooperaj) [inline includes/searcher/class-wp-lucene-search-searcher-azure.php]: Brackets!

File: includes/searcher/class-wp-lucene-search-searcher-azure.php [hunk 5]
```diff
@@ -28,15 +33,76 @@ protected function get_request_object( $body )
...
+
+    protected function parse_facets( $facets )
+    {
+        $facet_data = [];
+        foreach ( $facets as $key => $values ) {
+            $acf_field = acf_get_field( $key );
+            $pretty_post_types = acf_get_pretty_post_types( WP_Lucene_Search_Options_Helper::get_post_types() );
+            $facet_label = WP_Lucene_Search_Helper::get_nice_label( $key );
+            $facet_values = [];
+            foreach ( $values as $value ) {
```
Comment (cooperaj) [inline includes/searcher/class-wp-lucene-search-searcher-azure.php]: Break this up

File: public/class-wp-lucene-search-public.php [hunk 6]
```diff
@@ -104,11 +104,33 @@ public function override_search(WP_Query $query)
             return;
         }
 
+        $args['filter'] = isset($_GET['wls']) ? $_GET['wls'] : [];
```
Comment (cooperaj) [inline public/class-wp-lucene-search-public.php]: Breakt this up a bit

File: includes/searcher/class-wp-lucene-search-searcher-azure.php (hunk 4 above)
Comment (Hodgy) [reply includes/searcher/class-wp-lucene-search-searcher-azure.php]: Fixed

File: includes/searcher/class-wp-lucene-search-searcher-azure.php (hunk 3 above)
Comment (Hodgy) [reply includes/searcher/class-wp-lucene-search-searcher-azure.php]: Fixed

File: includes/class-wp-lucene-search-helper.php (hunk 2 above)
Comment (Hodgy) [reply includes/class-wp-lucene-search-helper.php]: Moved this into the searcher and split out some code into a new method

**Pull request merged**